internal/uri/uri.go

```diff
@@ -0,0 +1,71 @@
+package uri
+
+import (
+	"path/filepath"
+	"strings"
+)
+
+func (u Uri) IsLocal() bool {
+	if len(u.Scheme()) > 2 {
+		return false
+	} else {
+		return true
+	}
+}
+
+// actualUrl returns the url without the query string
+func (u Uri) ActualUrl() string {
+	terminator := strings.LastIndex(u.uri, ",")
+	if terminator != -1 {
+		return u.uri[0:terminator]
+	} else {
+		return u.uri
+	}
+}
+
+func (u Uri) Scheme() string {
+	return u.url.Scheme
+}
+
+func (u Uri) FileExtension() string {
+	_, f := filepath.Split(u.ActualUrl())
+	return strings.TrimLeft(filepath.Ext(f), ".")
+}
+
+func (u Uri) Filename() string {
+	filename := filepath.Base(u.ActualUrl())
+	if filename == "" {
+		panic("filename is empty")
```

Comment:
>	panic("filename is empty") [](http://example.com/codeflow?start=1&length=27)

this reliance on file name in the user's URL seems very limiting, even at this phase of our development. 
It's fine for the Uri to say "I don't know the file name" then let the consumer decide what to do. The consumer might say "ok let's download it and see what we get" or "fail if the user didn't tell us what mechanism to use".
Presumably if the mechanism is "restore" we can assume it's a "bak", correct?